ecs-agent/manageddaemon/managed_daemon.go

```diff
@@ -0,0 +1,326 @@
+// set required mounts
+// and override host paths in favor of agent defaults
+// when a duplicate SourceVolumeID is given, the last Mount wins
+func (md *ManagedDaemon) SetMountPoints(mountPoints []*MountPoint) error {
+	var mountPointMap = make(map[string]*MountPoint)
+	for _, mp := range mountPoints {
+		if mp.SourceVolumeID == defaultAgentCommunicationMount {
+			mp.SourceVolumeHostPath = fmt.Sprintf("%s/%s/", defaultAgentCommunicationPathHostRoot, md.imageName)
+			md.agentCommunicationMount = mp
+		} else if mp.SourceVolumeID == defaultApplicationLogMount {
+			mp.SourceVolumeHostPath = fmt.Sprintf("%s/%s/", defaultApplicationLogPathHostRoot, md.imageName)
+			md.applicationLogMount = mp
+		} else {
+			mountPointMap[mp.SourceVolumeID] = mp
+		}
+	}
+	mountResult := []*MountPoint{}
+	for _, mp := range mountPointMap {
+		mountResult = append(mountResult, mp)
+	}
+	md.mountPoints = mountResult
+	return nil
+}
+
+// Used to set or to update the agentCommunicationMount
+func (md *ManagedDaemon) SetAgentCommunicationMount(mp *MountPoint) error {
+	if mp.SourceVolumeID == defaultAgentCommunicationMount {
+		mp.SourceVolumeHostPath = fmt.Sprintf("%s/%s/", defaultAgentCommunicationPathHostRoot, md.imageName)
+		md.agentCommunicationMount = mp
+		return nil
+	} else {
+		return fmt.Errorf("AgentCommunicationMount %s must have a SourceVolumeID of %s", mp.SourceVolumeID, defaultAgentCommunicationMount)
+	}
+}
+
+// Used to set or to update the applicationLogMount
+func (md *ManagedDaemon) SetApplicationLogMount(mp *MountPoint) error {
+	if mp.SourceVolumeID == defaultApplicationLogMount {
+		mp.SourceVolumeHostPath = fmt.Sprintf("%s/%s/", defaultApplicationLogPathHostRoot, md.imageName)
+		md.applicationLogMount = mp
+		return nil
+	} else {
+		return fmt.Errorf("ApplicationLogMount %s must have a SourceVolumeID of %s", mp.SourceVolumeID, defaultApplicationLogMount)
+	}
+}
+
+func (md *ManagedDaemon) SetEnvironment(environment map[string]string) {
+	md.environment = make(map[string]string)
+	for key, val := range environment {
+		md.environment[key] = val
+	}
+}
+
+func (md *ManagedDaemon) SetLoadedDaemonImageRef(loadedImageRef string) {
+	md.loadedDaemonImageRef = loadedImageRef
+}
+
+// AddMountPoint will add by MountPoint.SourceVolume
+// which is unique to the task and is a required field
+// and will throw an error if an existing
+// MountPoint.SourceVolume is found
+func (md *ManagedDaemon) AddMountPoint(mp *MountPoint) error {
+	mountIndex := md.GetMountPointIndex(mp)
+	if mountIndex != -1 {
+		return fmt.Errorf("MountPoint already exists at index %d", mountIndex)
+	}
+	md.mountPoints = append(md.mountPoints, mp)
+	return nil
+}
+
+// UpdateMountPoint will update by
+// MountPoint.SourceVolume which is unique to the task
+// and will throw an error if the MountPoint.SourceVolume
+// is not found
+func (md *ManagedDaemon) UpdateMountPointBySourceVolume(mp *MountPoint) error {
+	mountIndex := md.GetMountPointIndex(mp)
+	if mountIndex < 0 {
```

Comment:
`mountIndex == -1` to assert that -1 indicates not found? same question for `DeleteMountPoint()`

Reply:
+1